### PR TITLE
feat: add schema and validator for UMB message content

### PR DESCRIPTION
**Pull request by goldmann**

### CODE REVIEW

File: service/src/main/resources/message-success-schema.json [hunk 1]
```diff
@@ -0,0 +1,89 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://jboss.org/sbomer/message.json",
+  "title": "UMB message",
+  "description": "A UMB message sent after the Sbom generation process is done",
```
Comment (mprpic) [inline service/src/main/resources/message-success-schema.json:5]:
```suggestion
  "description": "A UMB message sent after the SBOM generation process is done",
```

File: service/src/main/resources/message-success-schema.json [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    "productConfig": {
+      "type": "object",
+      "description": "Product information related to the Sbom and build",
+      "properties": {
+        "errataTool": {
+          "description": "Product information from the Errata Tool",
+          "type": "object",
+          "properties": {
+            "productName": {
+              "description": "Errata product name",
```
Comment (mprpic) [inline service/src/main/resources/message-success-schema.json:21]:
```suggestion
              "description": "Errata Tool product name",
```

File: service/src/main/resources/message-success-schema.json [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+              "type": "string"
+            },
+            "productVersion": {
+              "description": "Errata product version",
```
Comment (mprpic) [inline service/src/main/resources/message-success-schema.json:25]:
```suggestion
              "description": "Errata Tool product version",
```

File: service/src/main/resources/message-success-schema.json [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      "properties": {
+        "errataTool": {
+          "description": "Product information from the Errata Tool",
+          "type": "object",
+          "properties": {
+            "productName": {
+              "description": "Errata product name",
+              "type": "string"
+            },
+            "productVersion": {
```
Comment (mprpic) [inline service/src/main/resources/message-success-schema.json:24]: I think I missed this in the Google doc, but I think this should be a Product Variant. At least that is the object that CPEs are defined on in Errata Tool.

File: service/src/main/resources/message-success-schema.json [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    "purl": {
+      "description": "A package URL to uniquely identify the main component",
+      "type": "string"
+    },
+    "productConfig": {
+      "type": "object",
+      "description": "Product information related to the Sbom and build",
+      "properties": {
+        "errataTool": {
+          "description": "Product information from the Errata Tool",
```
Comment (mprpic) [inline service/src/main/resources/message-success-schema.json:17]:
```suggestion
          "description": "Product information from Errata Tool",
```

File: service/src/main/resources/message-success-schema.json (hunk 4 above)
Comment (goldmann) [reply service/src/main/resources/message-success-schema.json:24]: `productVariant` added.

File: service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                                .errataTool(
+                                        ErrataProductConfig.builder()
+                                                .productName("BLAH")
+                                                .productVersion("AABB")
+                                                .productVariant("DDDD")
+                                                .build())
+                                .build())
+                .sbom(
+                        Sbom.builder()
+                                .id("aaa")
```
Comment (vibe13) [inline service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java:56]: .id("429305915731435500")

File: service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                                .link("https://sbomer/api/v1alpha1/sboms/1234")
+                                .bom(
+                                        Bom.builder()
+                                                .format(BomFormat.CYCLONEDX)
+                                                .version("1.4")
+                                                .link("https://sbomer/api/v1alpha1/sboms/1234/bom")
```
Comment (vibe13) [inline service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java:62]: .link("https://sbomer/api/v1alpha1/sboms/429305915731435500/bom").
Please note that this endpoint does not exist, we need to implement it

File: service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                                                .build())
+                                .build())
+                .build(
+                        Build.builder()
+                                .id("AABBCC")
+                                .buildSystem(BuildSystem.PNC)
+                                .link("https://orch.pnc/builds/AABBCC")
```
Comment (vibe13) [inline service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java:69]: .link("https://orch.psi.redhat.com/pnc-rest/v2/builds/AWY5AHXBLAQAA")

File: service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                                .bom(
+                                        Bom.builder()
+                                                .format(BomFormat.CYCLONEDX)
+                                                .version("1.4")
+                                                .link("https://sbomer/api/v1alpha1/sboms/1234/bom")
+                                                .build())
+                                .build())
+                .build(
+                        Build.builder()
+                                .id("AABBCC")
```
Comment (vibe13) [inline service/src/test/java/org/jboss/sbomer/test/messaging/GenerationFinishedMessageBodyTest.java:67]: .id("AWY5AHXBLAQAA")

**Review by vibe13 — CHANGES_REQUESTED** [review]
Looks good to me! I have put a comment related to the configuration in the application.yaml, we might put other feature flags nearby the new one you created.
In the schema validator test, I have put real examples of values, an noticed that we actually miss the endpoint requested there (link to bom given an sbom id; we do have links to bom - base and enriched, I guess we will point to the enriched one - given a pnc build Id or a root purl though, should we use one of them instead?).
Finally, are we interested in sending the notification about start of the generation / processing of an sbom, after the receival of a build id via UMB? Or we just send the completed ones?

**Comment by goldmann** [comment]
I do definitely plan to have all the feature-flags in a single place. Maybe I'll remove it for now and we can work on establishing some rules for them. Or we can just improve as we go.

Regarding real-world values -- sure, I can add these, these weren't that important for the though :) Yes, we are missing the endpoint, but I guess I would like to take another round of the REST API review, I'm currently lost in the API, it's very verbose.

Question about when to send message is a valid one but I don't have an answer. I think it depends how are UMB topic is structured. Generally being more verbose doesn't hurt.

I'll apply changes you proposed shortly!

**Comment by goldmann** [comment]
I'm now in the process of updating the code with your comments and will add the REST endpoint as well.

**Comment by goldmann** [comment]
@vibe13 Changes applied, endpoint added!

**Review by vibe13 — APPROVED** [review]
Great, thanks Marek!

**Pull request merged**